src/emulation/promises.ts

```diff
@@ -718,16 +727,35 @@ export async function readdir(
 			entries.push(entry);
 		}
 	}
-	const values: (string | Dirent)[] = [];
+
+	const values: (string | Dirent | Buffer)[] = [];
 	for (const entry of entries) {
-		values.push(typeof options == 'object' && options?.withFileTypes ? new Dirent(entry, await stat(join(path, entry))) : entry);
+		const fullPath = join(path, entry);
+
+		const stats = options?.recursive || options?.withFileTypes ? await stat(fullPath) : null;
+		values.push(options?.withFileTypes ? new Dirent(entry, stats!) : entry);
```

Comment:
In the sync variant, we check to see if the encoding is a buffer type and if so convert to a buffer. You may want to consider adding to the ternary:

```ts
values.push(options?.withFileTypes ? new Dirent(entry, stats!) : options?.encoding == 'buffer' ? Buffer.from(entry) : entry);
```